Pre-release Item merge

Spartwo opened the pull request:
Changes discussed in Discord messaging.

- Gamedata naming and file conventions changed to align with the standard set by other larger mods
- Part Research and Cargo Compatibiity
- Scanner module and configs updated to reflect the agreed change to no longer replace science parts
- Inclusion of a baseline datalinking module in communications parts pending near future implementation, done for persist compatibility.
- Inclusion of savegame-tied settings for items of player or battle-based preferences. Currently a retreat override, and scanning range multiplers.
- Support for docking ports and radial decouplers as weapon seperators
- Miscellanious Fixes

Spartwo committed:
Fixes to config files

- Re-instated expected naming conventions
- Fixed folder names
- Specified part model titles
- Disabled yet unimplemented firing modes

Spartwo committed:
Patches fixes

There's a lot wrong here and it's too much to list.

Spartwo committed:
I don't have 8

Spartwo committed:
Re-instated Settings file

- Implementation is only in escape pod but will update as files are vetted

Spartwo committed:
Fixed Settings Connection Error

Forgot to re-import the folder through VS as simply undeleting it from the folder won't do.

Spartwo committed:
Fix Object Tracking Module

Version of script in release branch was from before the decision to make new parts as opposed to overwrite them.

80% of code was redundant, push contains the version from around early October.

Spartwo committed:
Merge pull request #1 from Halbann/main

Fix overlay visibility and missile control issues

Spartwo committed:
misc fixes

- Scanner attachment rules fix(I think this one's on me)
- Put refresh rate back into the gui elements for performance

Spartwo committed:
Utilities Fixes

- Organised utilities
- Brought over extended short name tags from original branch
- Replaced getdecouplerchildren with the current version from original branch
- Fixed misspelling of shortenname
- Updated decoupler designator to match original branch

Spartwo committed:
Weapons Based Fixes

- Reverted Missile reference finder(reason in code)
- Ported thrust vector utilities from old repository
- Reimplemented game settings based update intervals from old repository
- Re-added disallowed withdrawl gamesetting
- Added rocket support for docking port decouplers
- Added rocket support for angled decouplers
-- Rockets now aim based on expected flight direction
- Added shells of datalink modules for next block of work
- changed some variable name cases to camel from pascal(that's my fault)
- Added base datalink module to all comms parts for future compatibility

Spartwo committed:
Datalink fixes

no performance impact but if someone decided to compile I'd feel bad if it didn't work

Spartwo committed:
Lateral Rockets Edits

- I can see this functioning as intended but code somewhere looks to be resetting the part
- Works correctly in stacked arrangements
- Will return to the problem

Spartwo committed:
Requested Naming Adjustments

Spartwo committed:
Requested Modifications

- Removed decoupler types that were believed included intentionally but were just leftovers
- Changed retreat alloance term
- Temporarily disabled shipcontroller variable update interval

Spartwo committed:
Camel Breeding

- Changed intertim missile reference command to the lighterweight one until thrust based is done
- Removed Weapon tag requirement from decouplers in rockets and missiles pending automatic endcap detection
- Changed code case in completed scripts as was intended before pre-release

Spartwo committed:
Fix disparity between withdrawl allowances

Spartwo committed:
Datalink file movement

VS placed the datalink.cs file as a copy inside properties, relocated and made changes to adjust.

Halbann commented on GameData/KCS/Parts/ShipController/shipController.cfg:
```diff
@@ -33,6 +44,7 @@ PART
 		{
 			name = General
 			displayName = General
+			themeName = General
```
```suggestion

```

themeName asks the game to search for an existing theme. In this case the theme doesn't exist so it creates an error on startup.

Halbann commented on GameData/KCS/Parts/ShipController/shipController.cfg:
```diff
@@ -1,27 +1,38 @@
 PART
 {
-	name = shipController
+	name = KCSShipController
```
Will changing the name affect ships that already have the ship controller attached? To be expected in a beta but it might be a problem if the ships can't be loaded at all. Same goes for the weapon controller.

Halbann replied:
It does make ships that already have the parts unloadable, what do you want to do? Several people have already set up or built new ships with the mod and Kerosene is working on a cinematic, that's my bad. Maybe shipController, weaponController, escapePodController should be kept in as hidden parts? That or we have to keep using the generic names for these parts.

Spartwo replied:
naming retained to be safe

Halbann commented on GameData/KCS/Parts/WeaponController/weaponController.cfg:
```diff
@@ -0,0 +1,122 @@
+		{
+			name = Missile
+			displayName = Missile
+			themeName = Missile
```
```suggestion

```

Same reason as in the ship controller.

Halbann commented on GameData/KCS/Parts/WeaponController/weaponController.cfg:
```diff
@@ -0,0 +1,122 @@
+		{
+			name = Rocket
+			displayName = Rocket
+			themeName = Rocket
```
```suggestion

```

Halbann commented on GameData/KCS/Parts/WeaponController/weaponController.cfg:
```diff
@@ -0,0 +1,122 @@
+		{
+			name = Firework
+			displayName = Firework
+			themeName = Firework
```
```suggestion

```

Halbann commented on GameData/KCS/Patches/Scanners.cfg:
```diff
@@ -0,0 +1,29 @@
+  	}
+}
+
+//give all AI modules an inherit visual range
```
Clarification about what object tracking does at the moment.

```suggestion
// Give all AI modules an inherent lock range.
```

Halbann commented on GameData/KCS/Patches/Scanners.cfg:
```diff
@@ -0,0 +1,29 @@
+
```
I think the defaults for these values (and the value in smallScanner.cfg) could be increase by 5 or 10 times. 1000 m for the orbital scanner seems very short range in my opinion. If that's the go-to part for mid-sized ships then I think it should be at least 5 km.

Could you include mediumDishAntenna and HighGainAntenna5_v2 from ScanningParts.cfg in this repo? You could change the values to make them fit in below the smaller scanner as an improvement for very small ships. They're comms parts but I think the models make sense as radars.

Spartwo replied:
1000m with the default persist setting is 5000m

Halbann replied:
I see. Does the scaled value definitely show up on the info panel? If GetInfo gets called before OnStart it might not.

Spartwo replied:
It does, by default the small rotating scanner is 2500m and the stock one 5000m

Halbann replied:
Alright sounds good

Halbann commented on Source/KerbalCombatSystems/DataLink.cs:
```diff
@@ -0,0 +1,115 @@
+using System.Threading.Tasks;
+using KSP.UI.Screens;
+using UnityEngine;
+
```
How is this system going to work when it's finished? Can ModuleDataLink be removed? Do the power levels vary depending on the part?

Spartwo replied:
The module is applied to all parts that have the vanilla ksp communications module. The downside of this is that it's shared between relays and recievers. The basic moduledatalink will apply the other two based on part type in the same fashion as the weapon controllers.

The communications system itself will work in the same fashion as regular commnet for easy onboarding, both recievers and transmitters have a power level that varies by part and the square root of the distance will dictate if target lists can be successfully sent. The modules will just tell the shipcontroller about other scanning radii in the same way that baseline scanners just tell the ship how far it can see.

Halbann replied:
Sounds good. Wouldn't it be safer and more easily configurable to specify the correct modules for each part in module manager though?

Spartwo replied:
This method would provide modded part support and it's a long list to individually outline, including the inbuilt antennas in command parts.

Halbann replied:
Yeah good point

Halbann commented on Source/KerbalCombatSystems/Debug.cs:
```diff
@@ -12,13 +12,18 @@ public class KCSDebug : MonoBehaviour
         private static List<float> times;
+
+        //relavent game settings
+        private int refreshRate;
```
I don't think there's a benefit to using refresh rate in this file. I'd rather it repeat at a known rate over something set by the user, considering it doesn't affect performance.

Spartwo replied:
I believe this was an inherited block and most likely not relavant after end user linedraws got moved elsewhere so this whole performance draw is toggleable

Halbann commented on Source/KerbalCombatSystems/Debug.cs:
```diff
@@ -30,6 +35,7 @@ private void Update()
                 //switch bool return
                 showLines = !showLines;
 
+                //this block strikes me as redundant to line 91 but not actively harmful
```
```suggestion

```
It hides unused lines that would otherwise linger until they get hidden by the line cleaner.

Spartwo replied:
Understood

Halbann commented on Source/KerbalCombatSystems/DecouplerDesignator.cs:
```diff
@@ -2,17 +2,17 @@
             UI_ChooseOption(controlEnabled = true, affectSymCounterparts = UI_Scene.None,
-            options = new string[] { "Default", "Weapon", "Countermeasure", "Escape Pod", "Warhead" })]
-        public string DecouplerType = "Default";
+            options = new string[] { "Default","Escape Pod", "Countermeasure", "Warhead" })]
```
```suggestion
            options = new string[] { "Default", "Escape Pod", "Warhead" })]
```
Could we remove countermeasures here? Do you have plans to add them? In the mean time it could be confused to be related to interceptors.

Spartwo replied:
Will remove for time being, plans to include them for features like reactive armor or lore chaff

Halbann commented on Source/KerbalCombatSystems/KerbalCombatSystems.csproj:
```diff
@@ -9,7 +9,7 @@
     <RootNamespace>KerbalCombatSystems</RootNamespace>
     <AssemblyName>KerbalCombatSystems</AssemblyName>
-    <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
```suggestion
    <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
```
I upgraded to Visual Studio 2022 after reseting my computer and it doesn't support 4.6.1. BD uses 4.8 so it's definitely safe to use. We both have to be on the same version because we can't add this file to the gitignore.

Spartwo replied:
I reverted to 4.6.1 in the initial rush to make changes after the sudden release. I'll be ugrading versions going forward but couldn't leave it out of the pull.

Halbann commented on Source/KerbalCombatSystems/EscapePod.cs:
```diff
@@ -9,20 +9,24 @@ namespace KerbalCombatSystems
+        #region Fields
         const string EscapeGuidanceGroupName = "Escape Pod Guidance";
 
         private List<Part> AIPartList;
```
Could you replace AIPart with controller or shipController for clarity?

Halbann commented on Source/KerbalCombatSystems/MissileGuidance.cs:
```diff
@@ -69,12 +69,15 @@ private IEnumerator Launch()
                 seperator.Separate();
 
                 // I will think of something smarter in the future time.
+                // I do not agree with this inclusion, just let badly designed missiles explode - Spartwo
```
I'm pretty sure it doesn't work anyway so you can remove it if you like.

Spartwo replied:
Based and consequences pilled

Halbann commented on Source/KerbalCombatSystems/RocketTargeting.cs:
```diff
@@ -12,8 +12,9 @@ public class ModuleRocket : ModuleWeapon
-        internal ModuleDecouple decoupler;
+        private List<Seperator> decouplers;
+        internal Seperator decoupler;
+        Vector3 AimVector;
```
Please could you switch this to camel case.

Spartwo replied:
Are you referring to AimVector?

Halbann replied:
Yes

Halbann commented on Source/KerbalCombatSystems/RocketTargeting.cs:
```diff
@@ -150,16 +152,19 @@ public float Calculate()
             if (engines.Count < 1)
                 return -1;
 
-            float cosineLosses;
```
Rockets are usually placed at slight angle for spin which creates cosine losses. The predicted acceleration will be off for the whole simulation without accounting for cosine losses.

Spartwo replied:
If possible: review the generic utility to highlight where averaging the direction and magnitude of the forward thrust won't properly account for the cosine losses.

Halbann replied:
Total thrust for the same rocket over five updates:

Current:
100.7655
100.7653
100.7654
100.7655
100.7655

Suggested:
100.659
100.6579
100.6593
100.66
100.66

Not exactly the same but they all still hit and the new method accounts for engines with multiple thrust transforms so I'm happy to use it.

Spartwo replied:
Glad of it, utility based means we can apply it for other purposes

Halbann commented on Source/KerbalCombatSystems/RocketTargeting.cs:
```diff
@@ -259,14 +264,23 @@ private IEnumerator FireRocket()
+            AssignReference(decoupler.part.GetReferenceTransform(), AimVector);
+        }
+
+        private void AssignReference(Transform Position, Vector3 Direction)
```
Are you sure this works? I think this will just rotate the decoupler towards aimVector.

Spartwo replied:
It does neither at the moment and acts identically to initial beta release just using decoupler up vector based on tests. The goal is to change the decouplers part reference to align parallel with aim vector.

Halbann replied:
This function visibly rotates the decoupler. Transform is a reference type so when you change the rotation of the transform on line 282 you are rotating the decoupler. Transforms have to be attached to a GameObject so you'd need to create a new gameobject if you wanted a new transform to work with.

I'd suggest removing the call to this function for now and working out something else to get the ship to aim using the thrust vector, possibly after we have a custom PID (which will be some time) because then we can use quaternions to aim and won't need to set reference transforms.

Spartwo replied:
Version for now comments out the call, lack of function excludes a lot of different weapon and missile designs so is considered too high a priority to leave until custom PID.

Halbann commented on Source/KerbalCombatSystems/ShipController.cs:
```diff
@@ -14,9 +14,10 @@ public class ModuleShipController : PartModule
+        public float updateInterval;
         public float emergencyUpdateInterval = 0.5f;
+        public float combatUpdateInterval = 2.5f;
+        private bool allowWithdrawl;
```
I think it should be withdrawal rather than withdrawl

Spartwo replied:
Yea that's correct

Halbann commented on Source/KerbalCombatSystems/Settings.cs:
```diff
@@ -0,0 +1,95 @@
+        private int refreshRateDefault = 5;
+        //change inbuilt code to use the minimum functional value
+
+        [GameParameters.CustomParameterUI("Allow Withdrawl",
```
```suggestion
        [GameParameters.CustomParameterUI("Allow Withdrawal",
```
I think it should be withdrawal rather than withdrawl

Halbann commented on Source/KerbalCombatSystems/EscapePod.cs:
```diff
@@ -89,35 +120,14 @@ private IEnumerator RunEscapeSequence()
 
-            yield return new WaitForSeconds(5f);
+            yield return new WaitForSeconds(refreshRate);
             StartCoroutine(StatusRoutine());
```
You could make this more readable and possibly safer by using a while loop instead.

Halbann commented on Source/KerbalCombatSystems/Utils.cs:
```diff
@@ -6,56 +6,35 @@ namespace KerbalCombatSystems
 {
     public static partial class KCS
```
Please could you undo the function moves for now? Or give me a summary of what you changed? The diff is impossible to read. The original order was:

1 FindDecoupler
2 VesselDistance
3 GetMaxAcceleration
4 GetMaxThrust
5 RayIntersectsVessel
6 TargetLead
7 TryToggle
8 FindDecouplerChildren
9 FromTo
10 RelVel
11 AngularAcceleration
12 Integrate
13 SolveTime
14 SolveDistance
15 AngularVelocity
16 ShortenName
17 FindCommand
18 FindController
19 AveragedSize
20 FuelMass
21 DryMass

Spartwo replied:
No changes to the existing functions. Newer order is
# Get Properties
1 AveragedSize
2 Fue Mass
3 DryMass
4 GetMaxAcceleration
5 GetMaxThrust
6 GetFireVector
7 GetMeanVector(Possibly a poorly named function but gets the thrust direction for ksp parts that don't have it correlating to up)
# DoAction
8 ShortenName
9 TryToggle
# Finders
10 FindCommand
11 FindController
12 FindDecoupler
13 FindDecouplerChldren
# Physics Calcs
14 FromTo
15 RelVel
16 AngularAcceleration
17 AngularVelocity
18 Integrate
19 SolveTime
20 SolveDistance
21 TargetLead
22 VesselDistance
23 RayIntersectsVessel

Spartwo replied:
Issue was the difficulty in rolling it back successfully

Halbann replied:
I was able to see a proper diff by committing the function moves on my branch and it all looks good. It thinks the differences are merge conflicts now but I'll resolve them during the merge.

Spartwo committed:
Update GameData/KCS/Parts/ShipController/shipController.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update GameData/KCS/Parts/WeaponController/weaponController.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update GameData/KCS/Parts/WeaponController/weaponController.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update GameData/KCS/Parts/WeaponController/weaponController.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update GameData/KCS/Patches/Scanners.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update Debug.cs

Remove more "legacy" value from initial merge

Spartwo committed:
Update Debug.cs

Spartwo committed:
Update Source/KerbalCombatSystems/DecouplerDesignator.cs

Countermesure will be re-added at a later date when related features are pushed

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Withdrawl spelling correction

Spartwo committed:
Update Source/KerbalCombatSystems/KerbalCombatSystems.csproj

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Aimvector decamal

Would have gotten to it at some point

Spartwo committed:
Merge branch 'main' of https://github.com/Spartwo/StockCombatAI

Halbann commented on GameData/KCS/Parts/WeaponController/weaponController.cfg:
```diff
@@ -33,6 +44,7 @@ PART
 	{
 		name = ModulePartVariants
 		baseVariant = Missile
+		baseThemeName = Missile
```
I missed this one. It also throws an error.

Halbann replied:
```suggestion

```
This one also throws an error.

Halbann commented on GameData/KCS/Patches/Scanners.cfg:
```diff
@@ -0,0 +1,29 @@
+
+@PART[SurveyScanner]
```
```suggestion
// baseDetectionRange is modified by 5x by default.

@PART[SurveyScanner]
```

Halbann commented on Source/KerbalCombatSystems/EscapePod.cs:
```diff
@@ -44,16 +49,42 @@ public void ManualEscape(KSPActionParam param)
 
             // try to pop decoupler
-            if (seperator != null)
+            try
```
A failed try-catch here creates an error in the log. If the expectation is that it is acceptable for seperator to be null then I think it should be an if-then so we don't create an error. Pretty sure the code path would be the same but without the error.

Spartwo replied:
Concern is if the decoupler exists but already fired will it throw up an error with the current decouple subfunction

Spartwo committed:
Update GameData/KCS/Parts/WeaponController/weaponController.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update GameData/KCS/Patches/Scanners.cfg

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Bugfixes

- Rolled back already implemented names in terms of naming conventions due to concerns over craft compatibility
- rebuilt universal seperator system within the decoupler designator. Avoids memory leak from repeat child searches.
- fix for baseplates being decouplers
- fix for radials not being decouplers
- added support for ports as decouplers and for repeat action
- fixed lack of support for manual rocket fire

Spartwo commented:
Concern over the change to the decoupler designator and removal of "weapon" tag, will not cause craft incompatibility but will likely return invalid and broken values to anyone who has craft set to it. Would need to swap to "default" manually

EDIT: Retracted, reverts to default on loaded craft files.

Spartwo commented:
> Concern over the change to the decoupler designator and removal of "weapon" tag, will not cause craft incompatibility but will likely return invalid and broken values to anyone who has craft set to it. Would need to swap to "default" manually

Retracted, reverts to default on loaded craft files.

Halbann commented on Source/KerbalCombatSystems/DecouplerDesignator.cs:
```diff
@@ -1,18 +1,52 @@
             UI_ChooseOption(controlEnabled = true, affectSymCounterparts = UI_Scene.None,
-            options = new string[] { "Default", "Weapon", "Countermeasure", "Escape Pod", "Warhead" })]
-        public string DecouplerType = "Default";
+            options = new string[] { "Default", "Escape Pod", "Warhead" })] //re-add "Countermeasure" at later date
```
```suggestion
            options = new string[] { "Default", "Warhead", "Escape Pod" })] //re-add "Countermeasure" at later date
```

What do you think of ordering them like this instead?

Spartwo replied:
Makes no difference to me

Halbann commented on Source/KerbalCombatSystems/DecouplerDesignator.cs:
```diff
@@ -1,18 +1,52 @@
+                      guiActiveEditor = false,
+                      guiName = "test",
+                      groupName = DecouplerDesignationGroupName,
+                      groupDisplayName = DecouplerDesignationGroupName)]
```
Is this button visible in the right click menu? It would be a duplication of the existing buttons.

Spartwo replied:
Testing button, forgot to remove.

Spartwo committed:
Update Source/KerbalCombatSystems/DecouplerDesignator.cs

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Remove temporary test visual

Spartwo committed:
Minor fix

root part was being inserted into the end of the decoupler list, moved to start

Halbann commented on Source/KerbalCombatSystems/RocketTargeting.cs:
```diff
@@ -84,7 +86,7 @@ public override Vector3 Aim()
             float aimTolerance = Vector3.Angle(targetVector, targetVector + targetRadius);
 
-            bool onTarget = Vector3.Angle(leadVector.normalized, decoupler.transform.up) < aimTolerance;
+            bool onTarget = Vector3.Angle(leadVector.normalized, aimVector) < aimTolerance;
```
```suggestion
            bool onTarget = Vector3.Angle(leadVector.normalized, decoupler.transform.up) < aimTolerance;
```
Because the ship isn't currently aiming with aimVector, onTarget will never be true unless the difference between aimVector and the current reference transform is less than aimTolerance. This should probably just be reverted until your re-organisation of rocket targeting.

Halbann committed:
Ship controller fix

updateInterval needs to be initialised at start

Spartwo committed:
Update Source/KerbalCombatSystems/RocketTargeting.cs

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Halbann commented on Source/KerbalCombatSystems/DecouplerDesignator.cs:
```diff
@@ -1,18 +1,47 @@
+        [KSPField(isPersistant = true)]
+            public bool seperated = false;
+
+        const string DecouplerDesignationGroupName = "Seperator Designation";
```
```suggestion
        const string DecouplerDesignationGroupName = "KCS Designation";
```

What do you think of this rename? I'm not sure myself. I think it would be good to let the user know where the extra module is coming from though, otherwise I don't think it's immediately obvious what it's for.

![image](https://user-images.githubusercontent.com/30965946/204271593-02031bc0-2c70-45cc-9ec3-d57f66962631.png)

Spartwo replied:
Saying separator twice is a bit redundant in fairness

Halbann commented on Source/KerbalCombatSystems/RocketTargeting.cs:
```diff
@@ -225,41 +220,43 @@ public float Calculate()
                 //check for engine modules on the part and stop if not found
-                if (Module == null) continue;
+                if (module == null) continue;
+                Debug.Log("not null");
```
```suggestion

```

Spartwo replied:
Testing leftover

Halbann commented on Source/KerbalCombatSystems/ShipController.cs:
```diff
@@ -300,7 +304,7 @@ private IEnumerator UpdateBehaviour()
 
             // Movement.
-            if (hasPropulsion && !hasWeapons && CheckWithdraw())
+            if (hasPropulsion && !hasWeapons && CheckWithdraw() && allowWithdrawal)
```
```suggestion
            if (allowWithdrawal && hasPropulsion && !hasWeapons && CheckWithdraw())
```

This would save having to do the vector maths in CheckWithdraw if allowWithdrawal is false.

Spartwo committed:
Update Source/KerbalCombatSystems/ShipController.cs

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update Source/KerbalCombatSystems/RocketTargeting.cs

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Spartwo committed:
Update Source/KerbalCombatSystems/DecouplerDesignator.cs

Co-authored-by: Lawrence <30965946+Halbann@users.noreply.github.com>

Halbann committed:
Merge branch 'main' into main
